### PR TITLE
device-common-install: add device hw validation tool

### PR DESCRIPTION
**Pull request by mtetreault**
Tools that allow to validate hardware specs. (HDD + RAM)

### CODE REVIEW

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  hdd_upper_bound=$(expr "$1" + "$2")
+  hdd_size="$(lsblk /dev/sda | head -n 2 | tail -n 1 | awk '{print $4}' | sed 's/[A-Z]//g')"
+
+  test "$hdd_size" -ge "$hdd_lower_bound" && test "$hdd_size" -le "$hdd_upper_bound"
+}
+
+is_expected_ram() {
+  # $1 = Expected value, $2 = Tolerance
+  if [ $# -lt 2 ]; then
+    return 1
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Error message to stderr required here. (e.g.: 1>&2 echo "ERROR: My error!"). Why not instead take the arguments the following way instead:

```bash
local expected_value
expected_value="${1?ERROR: My error message}"
```

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
+is_expected_hdd() {
+  # $1 = Expected value, $2 = Tolerance
+  if [ $# -lt 2 ]; then
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Same as below.

**Comment by mtetreault** [comment]
J'ai fait quelques changement pour limiter la duplication de code

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  device_mac=$(ifconfig "$network_interface" | grep ether | awk '{print $2}')
+
+  if [ null == "$(echo "$hw_config" | yq --arg key "$device_mac" '.[$key]')" ]; then
+    key="default"
+  else
+    key="$device_mac"
+  fi
+
+  expected_value="$(echo "$hw_config" | yq --arg key "$device_mac" \
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: When you read a value back to bash, use `-j`:


### Reading a value

As json:

```bash
$ echo "{}" | jq '.my_key.sub_key = "foo"' | jq '.my_key.sub_key'
"foo"
```

Shell ready, use the `-j` flag:

```bash
$ echo "{}" | jq '.my_key.sub_key = "foo"' | jq -j '.my_key.sub_key'
foo
```

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                                      --arg dev "$device" \
+                                      '.[$key][$dev].tolerance')"
+
+  if [ null == $expected_value ] || [ null == $tolerance ];then
+    1>&2 printf "ERROR: Could not read specification from \n $hw_config \n"
+    return 1
+  fi
+
+  if [ "$device" == ram ]; then
+    if ! is_expected_ram $expected_value $tolerance; then
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Always quote expansions.

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      1>&2 echo "ERROR: $device size is out of range, expecting $expected_value"
+      return 1
+    fi
+  else
+    if ! is_expected_hdd $expected_value $tolerance; then
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Always quote expansions.

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      1>&2 echo "ERROR: $device size is out of range, expecting $expected_value"
+      return 1
+    fi
+  fi
+
+  return 0
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: This is implicit.

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 7]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
+
+is_expected_hdd() {
+  local expected_value
+  local tolerance
+
+  expected_value="${1?ERROR variables is unset}"
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Note that as this is more of an assertion (programmer error) than an actual possible error, you can also leave the message empty which should yield a similar stderr:

```bash
$ my_fn 
bash: 1: parameter null or not set
```

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  tolerance="${2?ERROR variable is unset}"
+
+  hdd_lower_bound=$(expr "$1" - "$2")
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Shouldn't this use bc for the same reason  as below?

Also, `"${expected_value} - ${tolerance}"`

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+is_expected_ram() {
+  local expected_value
+  local tolerance
+
+  expected_value="${1?ERROR variables is unset}"
+  tolerance="${2?ERROR variable is unset}"
+
+  ram_lower_bound=$(expr "$1" - "$2")
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Idem as above.

Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: missing local for the 3 below vars.

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  ram_upper_bound=$(expr "$1" + "$2")
+  ram_size="$(free -m --si | grep Mem: | awk '{print $2}')"
+
+  if [ 1 -eq $(echo "$ram_size"'>='"$ram_lower_bound" | bc -l) ] && \
```
Comment (jraygauthier) [inline scripts/device-common-install/libexec/device_hw_validation.sh]: Idem as above

**Comment by jraygauthier** [comment]
@mtetreault : This last forced push does not address the changes above. Is this an error?

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  local tolerance
+
+  expected_value="${1?ERROR variables is unset}"
+  tolerance="${2?ERROR variable is unset}"
+
+  hdd_lower_bound=$(expr "$1" - "$2")
+  hdd_upper_bound=$(expr "$1" + "$2")
+  hdd_size="$(lsblk /dev/sda | head -n 2 | tail -n 1 | awk '{print $4}' | sed 's/[A-Z]//g')"
+
+  if [ 1 -eq $(echo "$hdd_size"'>='"$hdd_lower_bound" | bc -l) ] && \
```
[the remark this answers is not in the document]
Comment (mtetreault) [reply scripts/device-common-install/libexec/device_hw_validation.sh]: done

File: scripts/device-common-install/libexec/device_hw_validation.sh [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  local module
+  local device
+
+  hw_config="${1?ERROR variables is unset}"
+  network_interface="${2?ERROR variable is unset}"
+  device="${3?ERROR variable is unset}"
+
+  device_mac=$(ifconfig "$network_interface" | grep ether | awk '{print $2}')
+
+  if [ null == "$(echo "$hw_config" | yq --arg key "$device_mac" '.[$key]')" ]; then
```
[the remark this answers is not in the document]
Comment (mtetreault) [reply scripts/device-common-install/libexec/device_hw_validation.sh]: Oh!, good to know!!

**Comment by jraygauthier** [comment]
@mtetreault : Still, you resolved without fixing the comments.

In particular, the comment about keeping `local`  keyword close to the first assignment.

**Comment by mtetreault** [comment]
> @mtetreault : Still, you resolved without fixing the comments.
> 
> In particular, the comment about keeping `local` keyword close to the first assignment.

I resolved the issue because they are resolved on my local branch. It has been pushed yet because I introduced a bug at the same time. I am working on them as we speak

**Comment by mtetreault** [comment]
Everything should be there

**Pull request merged**